metricbeat/module/windows/perfmon/reader.go

```diff
@@ -0,0 +1,172 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// +build windows
+
+package perfmon
+
+import (
+	"regexp"
+	"strconv"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+var (
+	processRegexp = regexp.MustCompile(`(.+?)#[1-9]+`)
+)
+
+// Reader will contain the config options
+type Reader struct {
+	query             Query             // PDH Query
+	instanceLabel     map[string]string // Mapping of counter path to key used for the label (e.g. processor.name)
+	measurement       map[string]string // Mapping of counter path to key used for the value (e.g. processor.cpu_time).
+	executed          bool              // Indicates if the query has been executed.
+	log               *logp.Logger      //
+	groupMeasurements bool              // Indicates if measurements with the same instance label should be sent in the same event
+}
+
+// NewReader creates a new instance of Reader.
+func NewReader(config Config) (*Reader, error) {
+	var query Query
+	if err := query.Open(); err != nil {
+		return nil, err
+	}
+
+	r := &Reader{
+		query:             query,
+		instanceLabel:     map[string]string{},
+		measurement:       map[string]string{},
+		log:               logp.NewLogger("perfmon"),
+		groupMeasurements: config.GroupMeasurements,
+	}
+	for _, counter := range config.CounterConfig {
+		childQueries, err := query.ExpandWildCardPath(counter.Query)
+		if err != nil {
+			if config.IgnoreNECounters {
+				switch err {
+				case PDH_CSTATUS_NO_COUNTER, PDH_CSTATUS_NO_COUNTERNAME,
+					PDH_CSTATUS_NO_INSTANCE, PDH_CSTATUS_NO_OBJECT:
+					r.log.Infow("Ignoring non existent counter", "error", err,
+						logp.Namespace("perfmon"), "query", counter.Query)
+					continue
+				}
+			} else {
+				query.Close()
+				return nil, errors.Wrapf(err, `failed to expand counter (query="%v")`, counter.Query)
+			}
+		}
+		if childQueries == nil || len(childQueries) == 0 {
+			return nil, errors.Wrapf(err, `failed to expand counter (query="%v")`, counter.Query)
```

Comment:
`query.Close()` here too, or use this [`cleanup`](https://github.com/elastic/beats/pull/12134) helper.